### ISSUE TITLE
Upstreaming `web-task`

### ISSUE THREAD

**Issue by NthTensor** [rustwasm/wasm-bindgen#5048]
I have written a [fork](https://crates.io/crates/web-task) of `wasm-bindgen-futures` that uses [`async-task`](https://crates.io/crates/async-task) for it's internal task model. I'm opening this issue to discuss the possibility of up-streaming it into the `wasm-bindgen` repo.

## Motivation

The need for this crate grew out of bevy's crossplatform async runtime. On native targets, bevy runs on top of executors that use `async-task`. These executors provide `spawn` and `spawn_local` functions that return `async_task::Task<T>`. Tasks are themselves futures, and also can be used to cancel the spawned future,or to await it's result.

On web-targets, however, bevy instead uses the `spawn_local` function from `wasm_bindgen_futures`. This does not return a task, does not support cancellation, and only accepts futures that output `()`. To get cross-platform code compiling, we have had to jerry-rig a mock of the `Task` api, which provides empty methods like `cancel()` and `detach()`. Our task wrapper has been an endless source of bugs and, even with it in place, writing cross-platform code remains quite difficult. 

## Goals

Our primary goal was simply to gain feature parity with `spawn` and `spawn_local` provided by executors like `smol`, but on the web. We want to:
1. Get a task when we spawn something.
2. Be able to cancel the task (by calling `.cancel()` or dropping the task).
3. Be able to `.detach()` the task (which makes things perform as they do now, in the background).
4. Be able to get spawn futures with nontrivial outputs, and be able to `.await` tasks to retrieve those outputs.

A secondary goal was to provide a way to easily convert a non-send future (like the ones created from promises) into a send one, by mapping the output to a send value and spawning it to get a send task.
```rust
let task: Task<MySendType> = spawn_local(async {
    let js_result = nonsend_browser_api().await;
    MySendType::from(js_result)
});
```

These are both important for bevy's wider ambitions of full runtime parallelism on the web via web-workers. I imagine they will be salient to many other web-focused rust projects as well.

## Scope

The crate provides two functions: `spawn` and `spawn_local`. Both return `Task<T>` and are generally as one would expect. None of the other promises-related stuff has been copped from `wasm-bindgen-futures`. The legacy support code has also been omitted, but could be easily re-incorporated. It was pretty trivial to drop the abstractions from `async-task` into the existing single-threaded code. The multi threaded version is more complex and uses more unsafe, but is still only about 300 lines (with significant comments).

Now that the crate is complete, we have a few options. Bevy is considering officially adopting and maintain it. But since we'd be moving away from the defacto-standard ecosystem crate, maintained by hundreds of people and used by many thousands, and we figured we can't be the only people that want this, we'd rather find a way to give it more broad community ownership.

I've been given to understand that, while `wasm-bindgen` is not interested in changes to the futures crate (which is fair, it would be crazy to break that api), you might consider up-streaming other relevant crates. How do we feel about this one?

**Comment by guybedford** [comment]
Thanks for sharing the suggestion - I'd be happy to review and land an upstreaming if we gate it behind a feature like `async-task` and add it as a new `spawn_task` function alongside the existing one and integrate it into the same model, if something like that makes sense?

I also just landed https://github.com/wasm-bindgen/wasm-bindgen/pull/5049 today which you may be interested in as well.

**Comment by NthTensor** [comment]
https://github.com/wasm-bindgen/wasm-bindgen/pull/5049 looks really interesting!

I'll have to think about if/how this could be exposed as a feature without messing up the existing stuff. The only common point is really the queue, which is largely unchanged, so it could work. Very pleased to hear you are open to the idea.

Some of the other bevy devs are still auditing the implementation. Once that's done, I can have a go at a feature-gated reintegration PR. I guess it would go in `js-sys` now?

**Comment by alice-i-cecile** [comment]
I'm merging https://github.com/bevyengine/bevy/pull/21795 now; there should be plenty of stress-testing from our users in the coming months :)

**Comment by guybedford** [comment]
Happy to review any PRs here further, and yes it would go in js-sys at this point.
